lib/Service/SpeechToText/SpeechToTextService.php

```diff
@@ -0,0 +1,199 @@
+
+		$sttFolderName = $this->config->getAppValue(Application::APP_ID, 'stt_folder', '(not set)');
+		if ($sttFolderName === '(not set)') {
+			$sttFolderName = Application::SPEECH_TO_TEXT_REC_FOLDER;
+
+			if ($userFolder->nodeExists($sttFolderName)) {
+				$sttFolder = $this->getUniqueNamedFolder($userId);
+				$sttFolderName = $sttFolder->getName();
+			} else {
+				$sttFolder = $userFolder->newFolder($sttFolderName);
+			}
+			$this->config->setAppValue(Application::APP_ID, 'stt_folder', $sttFolderName);
+		} else {
+			$sttFolder = $userFolder->get($sttFolderName);
+			if (!$sttFolder instanceof Folder) {
+				// the folder created by this app was tampered with
+				// create a new one
+				$sttFolder = $this->getUniqueNamedFolder($userId);
+				$sttFolderName = $sttFolder->getName();
+				$this->config->setAppValue(Application::APP_ID, 'stt_folder', $sttFolderName);
+			}
+		}
+
+		$filename = (new DateTime())->format('d-M-Y-Hisu') . '.mp3';
+		$audioFile = $sttFolder->newFile($filename, fopen($tempFileLocation, 'rb'));
+
+		return $audioFile;
+	}
+
+	/**
+	 * @param string $userId
+	 * @param integer $try
+	 * @return Folder
+	 * @throws RuntimeException
+	 * @throws NotPermittedException
+	 */
+	private function getUniqueNamedFolder(string $userId, int $try = 3): Folder {
+		$userFolder = $this->rootFolder->getUserFolder($userId);
+		$sttFolderPath = Application::SPEECH_TO_TEXT_REC_FOLDER . ' ' . strval(4 - $try);
+
+		if ($userFolder->nodeExists($sttFolderPath)) {
+			if ($try === 0) {
+				// give up
+				throw new RuntimeException('Could not create a folder with a unique name');
+			}
+			return $this->getUniqueNamedFolder($userId, $try - 1);
+		}
+
+		return $userFolder->newFolder($sttFolderPath);
+	}
+
+	/**
+	 * Send transcription result notification
+	 * @param string $userId
+	 * @param string $result
+	 * @param boolean $success
+	 * @param int $taskType
+	 * @return void
+	 * @throws \InvalidArgumentException
+	 */
+	public function sendSpeechToTextNotification(string $userId, string $result, bool $success): void {
+		$manager = $this->notificationManager;
+		$notification = $manager->createNotification();
+
+		try {
+			$transcriptEntity = new Transcript();
+			$transcriptEntity->setUserId($userId);
+			$transcriptEntity->setTranscript($result);
+			// never seen transcripts should also be deleted in the cleanup job
+			$transcriptEntity->setLastAccessed(new DateTime());
+			$transcriptEntity = $this->transcriptMapper->insert($transcriptEntity);
+
+			$id = $transcriptEntity->getId();
+		} catch (\OCP\Db\Exception $e) {
+			$this->logger->error('Failed to save transcript in DB: ' . $e->getMessage());
+			$success = false;
+		}
+
+		$params = [
+			'appId' => Application::APP_ID,
+			'taskType' => Application::TASK_TYPE_SPEECH_TO_TEXT,
+			'result' => $result,
+			'target' => $this->urlGenerator->linkToRouteAbsolute(Application::APP_ID . '.SpeechToText.getResultPage', ['id' => $id])
```

Comment:
`$id` is not defined in case of error. Let's set it to 0 in the previous catch block (catching `\OCP\Db\Exception`).